### PR TITLE
Add clickable line numbers to console view

### LINKED ISSUE #43: Add line number to console output
Blue ocean implements this mostly with a css counter

image

### PR DESCRIPTION
**Pull request by timbrown5**
The aim was to add clickable line numbers to the console view (#43), but I hit some issues that required a refactor.

Fixes #43 

Changes:
- All API calls happen in PipelineConsole.
-- Data is passed down to children (they don't call API).
- DataTreeView no longer has state (all state is in PipelineConsole).
- Added (and switch to) function to get all steps in one go.
-- Reduces the number of API calls required.
-- Reduces the number of renders on startup.
-- Getting a list of steps no longer requires knowledge of the stages (these can be fetched asynchronously). 
- Moved state setup to the constructor
-- Reduces number of renders on startup.
-- Downside: UI might appear blank until API calls return.
- Added a new element to ConsoleLine to house line number
-- It's worth considering if we want the API to be a set of lines, instead of a wall of text. This would allow getting x lines to start with and having an element to load the next x lines.
--- x could be configurable - e.g. dropdown menu to select getting of 50, 100, 200 lines per request.
-- Lines could be fetched it batches up to a limit (e.g. 10 at a time until there are 100). 
- Added some debug logging - happy to remove it if not useful to others/slowing UI.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira

### CODE REVIEW

**Comment by timbrown5** [comment]
Screenshot 2022-09-15 at 17 20 45
Looking at it now, I'm not sure if some of the formatting has regressed (maybe the new layout has the wrong classes set?).

**Comment by timja** [comment]
What I'm seeing doesn't seem to match your screenshot:

image

Lots of spacing between log lines, and the line numbers look bad in dark theme.

Linking to log line doesn't seem to be working for me, if I link to a log line near the bottom of the page it doesn't scroll at all

**Comment by timja** [comment]
it seems to add an extra line number at the end?

image

File: src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx [hunk 1]
```diff
@@ -76,23 +45,37 @@ const getTreeItemsFromStepList = (stepsItems: StepInfo[]) => {
...
-      let stepsItems = getTreeItemsFromStepList(steps);
+    var i = stepsCopy.length
+    while (i--) {
+      let step = stepsCopy[i];
+      if (step.stageId == String(stageItemData.id)) {
+        stageSteps.push(step);
+        // Remove step from local copy - can only have one parent.
+        // This should reduce the total number of loops required.
+        stepsCopy.splice(i, 1);
+      }
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:75]: not sure if this is related but it would be good if nested parents worked right,

e.g. parallel(linux-11) -> stage(checkout(linux-11))

for an example of where it doesn't work see:
https://ci.jenkins.io/job/Plugins/job/slack-plugin/job/master/325/pipeline-console/?selected-node=25

Comment (timbrown5) [reply src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:75]: I'm not sure I follow this.
This should only be affecting steps, and (I think) steps should only have one parent.

It can be removed - it's an optimisation, not a functional component.

**Comment by timbrown5** [comment]
I will have a look - I think I haven't rebased to the correct branch locally.

Also, very happy to tweak the CSS, but it's not something I'm overly familiar with 😅.

File: src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx (hunk 1 above)
Comment (timja) [reply src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:75]: (this was broken before, wasn't sure if this code was related, I can raise a separate issue for it anyway)

**Comment by timja** [comment]
> I will have a look - I think I haven't rebased to the correct branch locally.
> 
> Also, very happy to tweak the CSS, but it's not something I'm overly familiar with 😅.

once it's less broken I can fiddle with CSS if needed no worries

**Comment by timbrown5** [comment]
Ahh, seems it's a change in Jenkins versions that causes the difference. updating to 2.346.1 I see this:
Looks like I might need to change the structure of `ConsoleLine` (adding a new `p` for each line is breaking the newer formatting.

**Comment by timja** [comment]
Should probably use spans, see https://weekly.ci.jenkins.io/job/Demo/7/console

**Comment by timbrown5** [comment]
I have updated it to work with newer Jenkins and dark mode. I have also added an offset to anchors so they don't disappear behind the breadcrumbs.

There is still an issue that anchors do not work when opting for the URL (if you reload the page they seem to). I think this is that the anchor gets evaluated before the item is rendered - but not sure how to fix it.

Jenkins 2.346.1
Screenshot 2022-09-16 at 10 00 16

Jenkins 2.332.4 Dark
Screenshot 2022-09-16 at 10 24 07

Jenkins 2.332.4 Default
Screenshot 2022-09-16 at 10 25 11

**Comment by timja** [comment]
lint is failing (`mvn spotless:apply`)

File: src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java [hunk 2]
```diff
@@ -42,12 +42,8 @@ public String getUrlName() {
     return "pipeline-console";
   }
 
-  // Consider making this return all steps (don't accept the node)
-  // If the steps include the parent ID (or be indexed by it), then I can save a list of steps to
-  // the
-  // React component state. Then I can cross-reference this when adding the child nodes instead of
-  // making a fetch call - which seems impossible to wait for the result of (maybe fetch is the
-  // wrong thing to use?)
+  // Legacy - leave incase we want to update a sub section of steps (e.g. if a stage if still
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java:45]:
```suggestion
  // Legacy - leave in case we want to update a sub section of steps (e.g. if a stage is still
```

**Comment by timja** [comment]
spotbugs failure

File: src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx (hunk 1 above)
Comment (timja) [reply src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:74]: raised https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/92

File: src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx [hunk 3]
```diff
@@ -27,47 +21,21 @@ export interface StepInfo {
...
-      },
-    },
-    // TODO: Make line show status of block (green = passed, red = failed, etc.)
-    group: {
-      marginLeft: 7,
-      paddingLeft: 18,
-      borderLeft: `1px dashed ${fade(theme.palette.text.primary, 0.4)}`,
-    },
-    label: {
-      fontStyle: (props: StageTreeItemProps) =>
```
Comment (mikedld) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:57]: Maybe I don't understand something but why was this removed? Okay if you don't want synthetic stage labels italicised, but if so they shouldn't probably be italicised in the graph view either...

Comment (timbrown5) [reply src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:57]: Ah, sorry. I removed the withStyle blocks because I though it was all handled in the scss.

**Comment by timbrown5** [comment]
Nice catch though.
I'm away for a week now, so I will have to revisit this when I'm back.

**Comment by timja** [comment]
Issues so far:

1. Line numbers with wrapping look a bit weird
image
2. I think a bit of padding between line numbers and content would be good
3. The page lost some width between the main branch, it was made quite wide to make more room for logs, intentional?
4. Linking to line number is a bit hit or miss, doesn't work on the pipeline from https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/92 (not a blocker)
5. Extra line at the end which looks a bit weird (blue ocean doesn't have this extra line)
6. CI build is failing

**Comment by timbrown5** [comment]
@timja thanks for the list. I feel I should have caught some of these things 😓.
I'm posting some links to remind help me when I get back (as I'm sure I will have forgotten everything by then 😅).
https://stackoverflow.com/questions/40842277/create-line-numbers-on-pre-with-css-only#40842325 - counts as CSS (can they be clickable anchors? Maybe with offset to avoid needing component to be loaded).
This might fix a few of the issues above 
If not https://css-tricks.com/multi-line-padded-text/

Is 3 a change in the default width of the tree view?

**Comment by timja** [comment]
> Is 3 a change in the default width of the tree view?
It's changed vs the main branch, I assume some CSS is missing now or something

**Comment by timbrown5** [comment]
>> Is 3 a change in the default width of the tree view?
> It's changed vs the main branch, I assume some CSS is missing now or something
I've had a look at the main branch locally and don't see the difference:

`jenkins-ci/main` [f8a7e8cfaa34c9eb7f5eb9235a716cd82b0a55fc]
Screenshot 2022-09-23 at 11 27 03
`console-line-numbers` [90017e2d8ecc8fd99861ef67d2ffb900cb24d307]
Screenshot 2022-09-23 at 11 21 07

That said I have rebased, so maybe that has fixed it?
I also added more padding (should be the same amount as Blue Ocean).

Screenshot 2022-09-23 at 10 51 00

**Comment by timja** [comment]
looks nice from screenshots.

would be good to get rid of that trailing line if possible

image

**Comment by timbrown5** [comment]
The only outstanding issue should be text wrapping (which I forgot to look at).

File: src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx [hunk 4]
```diff
@@ -69,23 +68,37 @@ const getTreeItemsFromStepList = (stepsItems: StepInfo[]) => {
         key={stepItemData.id}
         nodeId={String(stepItemData.id)}
         label={stepItemData.name}
+        stageId={String(stepItemData.stageId)}
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:71]: there's an error in the browser console that is likely this line (although could be another)

image

File: src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java [hunk 5]
```diff
@@ -67,12 +63,24 @@ public void getSteps(StaplerRequest req, StaplerResponse rsp) throws IOException
...
+  // - remove dependency of getting list of stages in frontend.
+  @WebMethod(name = "allSteps")
+  public void getAllSteps(StaplerRequest req, StaplerResponse rsp) throws IOException {
+    ObjectMapper mapper = new ObjectMapper();
+    PipelineStepList steps = stepApi.getAllSteps();
+    String stepsJson = mapper.writeValueAsString(steps);
+    if (logger.isDebugEnabled()) {
+      logger.debug("Steps: '" + stepsJson + "'.");
+    }
+    rsp.getWriter().append(stepsJson);
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java:81]: Content-Type header should be added to the response

**Review by timja — CHANGES_REQUESTED** [review]
ANSI color highlighting is broken:

image


main branch:

image

Also steps are out of order, build should be first but archive is.

This is built from the multi-branch job at:
https://github.com/timja-org/junit-attachments-test/tree/master

**Comment by timbrown5** [comment]
The refactor also seems to fix some of the performance issues (at least locally). I think this is due to allowing the fetch calls to be asynchronous.

Here are some charts from firefox (using the pipeline [from here](https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/39#issuecomment-891075485)).

Before:
Screenshot 2022-09-27 at 09 59 20

After:
Screenshot 2022-09-27 at 09 59 31

**Comment by timbrown5** [comment]
Fixed ANSI colours:
Screenshot 2022-09-27 at 13 37 21

Tested with:
```
pipeline {
    agent any
    options {
        ansiColor('xterm')
    }
    stages {
        stage('Build') {
            steps {
                echo '\033[34mHello\033[0m \033[33mcolorful\033[0m \033[35mworld!\033[0m'
            }
        }
    }
}
```

File: src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java (hunk 5 above)
Comment (timbrown5) [reply src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java]: Have switched to using HttpResponses.okJSON/errorJSON.

**Comment by timja** [comment]
Currently getting in browser console:

Page load:
image

When clicking on a step:
image


same pipeline as before
https://github.com/timja-org/junit-attachments-test/tree/master?rgh-link-date=2022-09-26T21%3A31%3A53Z

**Comment by timja** [comment]
All other issues seem fixed that I found, nice work!

**Comment by timbrown5** [comment]
> Currently getting in browser console:
> 
> Page load: image
> 
> When clicking on a step: image
> 
> same pipeline as before https://github.com/timja-org/junit-attachments-test/tree/master?rgh-link-date=2022-09-26T21%3A31%3A53Z

TIL: `npm run build:dev`

**Review by timja — APPROVED** [review]
LGTM, one non-blocking suggestion would be should clicking on a line highlight it? / hovering over the line number highlight the line?
